File: 백준/Gold/25189. 시니컬한 개구리/시니컬한 개구리.py
from collections import deque, namedtuple
from pprint import pprint

def printBoard(board):
    for y in range(1,N+1):
        for x in range(1, M+1):
            print(f"{board[y][x]:3d}",end=' ')
        print()

dy = [-1,0,1,0]
dx = [0,-1,0,1]
def OOB(y, x):
    return y<= 0 or y > N or x<=0 or x>M

def getMaxDistance(y, x, dir):
    if dir == 0:
        return y - 1
    elif dir == 1:
        return x - 1
    elif dir == 2:
        return N - y
    return M - x

if __name__=='__main__':
    N, M = map(int, input().split())
    sy,sx,ey,ex = map(int,input().split())
    board = [[0 for _ in range(M+1)] for _ in range(N+1)]
    for y in range(1,N+1):
        line = list(map(int,input().split()))
        for x in range(1,M+1):
            board[y][x] = line[x-1]

    INT_MAX = float('inf')
    dist = [[[INT_MAX for _ in range(M+1)] for _ in range(N+1)] for _ in range(2)]

    State = namedtuple('State', ['y','x','limit'])
    q = deque([State(sy,sx,0)])
    dist[0][sy][sx] = 0
    y_visited = [False for _ in range(N+1)]
    x_visited = [False for _ in range(M+1)]

    while q:
        cur = q.popleft()
        if(cur.y == ey and cur.x == ex):
            print(dist[cur.limit][cur.y][cur.x])
            exit()

        # 점프 X
        distance = board[cur.y][cur.x]
        for dir in range(4):
            ny = cur.y + dy[dir] * distance
            nx = cur.x + dx[dir] * distance
            if OOB(ny, nx) or dist[cur.limit][ny][nx] != INT_MAX:
                continue
            dist[cur.limit][ny][nx] = dist[cur.limit][cur.y][cur.x] + 1
            q.append(State(ny, nx, cur.limit))

        if cur.limit == 0:
            if not y_visited[cur.y]:
                y_visited[cur.y] = True
                ny = cur.y
                for x in range(1,M+1):
                    nx = x
                    if dist[1][ny][nx] == INT_MAX:
                        dist[1][ny][nx] = dist[0][cur.y][cur.x] + 1
                        q.append(State(ny,nx,1))

            if not x_visited[cur.x]:
                x_visited[cur.x] = True
                nx = cur.x
                for y in range(1, N+1):
                    ny = y
                    if dist[1][ny][nx] == INT_MAX:
                        dist[1][ny][nx] = dist[0][cur.y][cur.x] + 1
                        q.append(State(ny,nx,1))
            # # 점프 O
            # for dir in range(4):
            #     maxDistance = getMaxDistance(cur.y, cur.x, dir)
            #     for v in range(1, maxDistance + 1):
            #         ny = cur.y + dy[dir] * v
            #         nx = cur.x + dx[dir] * v
            #         if OOB(ny,nx) or dist[1][ny][nx] != INT_MAX:
            #             continue
            #         dist[1][ny][nx] = dist[0][cur.y][cur.x] + 1
            #         q.append(State(ny,nx,1))

    print(-1)

File: 백준/Gold/25189. 시니컬한 개구리/시니컬한 개구리.java
import java.util.*;
import java.io.*;

public class Main {
 
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	
    	st = new StringTokenizer(br.readLine());
    	sy = Integer.parseInt(st.nextToken());
    	sx = Integer.parseInt(st.nextToken());
    	ey = Integer.parseInt(st.nextToken());
    	ex = Integer.parseInt(st.nextToken());
    	
    	board = new int[N+1][M+1];
    	dist = new int[2][N+1][M+1];
    	for(int y=1; y<=N; y++) {
    		st = new StringTokenizer(br.readLine());
    		for(int x=1; x<=M; x++) {
    			board[y][x] = Integer.parseInt(st.nextToken());
    			for(int i = 0; i < 2; i++) {
    				dist[i][y][x] = INF;
    			}
    		}
    	}
    	
    	Queue<Tuple> q = new ArrayDeque<>();
    	dist[0][sy][sx] = 0;
    	q.add(new Tuple(sy,sx,0));
    	yVisited = new boolean[N+1];
    	xVisited = new boolean[M+1];
    	while(!q.isEmpty()) {
    		Tuple cur = q.poll();
    		
    		if(cur.y == ey && cur.x == ex) {
    			System.out.println(dist[cur.cnt][cur.y][cur.x]);
    			System.exit(0);
    		}
    		
    		int value = board[cur.y][cur.x];
    		for(int dir = 0; dir < 4; dir++) {
    			int ny = cur.y + dy[dir] * value;
    			int nx = cur.x + dx[dir] * value;
    			
    			if(OOB(ny,nx) || dist[cur.cnt][ny][nx] != INF)
    				continue;
    			dist[cur.cnt][ny][nx] = dist[cur.cnt][cur.y][cur.x] + 1;
    			q.add(new Tuple(ny,nx,cur.cnt));
    		}
    		
    		
    		if(cur.cnt == 0) {
    			if(!yVisited[cur.y]) {
    				yVisited[cur.y] = true;
    				int ny = cur.y;
    				
    				for(int nx=1; nx<=M; nx++) {
    					if(dist[1][ny][nx] == INF) {
    						dist[1][ny][nx] = dist[0][cur.y][cur.x] + 1;
    						q.add(new Tuple(ny,nx,1));
    					}
    				}
    			}
    			
    			if(!xVisited[cur.x]) {
    				xVisited[cur.x] = true;
    				int nx = cur.x;
    				
    				for(int ny = 1; ny<=N; ny++) {
    					if(dist[1][ny][nx] == INF) {
    						dist[1][ny][nx] = dist[0][cur.y][cur.x] + 1;
    						q.add(new Tuple(ny,nx,1));
    					}
    				}
    			}
    			
    		}
    		
    	}
    	System.out.println(-1);
    }
    static boolean yVisited[], xVisited[];
    static int dy[] = {-1,1,0,0};
    static int dx[] = {0,0,-1,1};
    static boolean OOB(int y,int x) {
    	return y<=0 || y>N || x<=0 || x>M;
    }
    
    static int N,M;
    static int INF = Integer.MAX_VALUE / 2;
    static StringTokenizer st;
    static int sy,sx,ey,ex;
    static int board[][];
    static int dist[][][];
    static class Tuple{
    	int y,x,cnt;
    	public Tuple(int y,int x,int cnt) {
    		this.y = y;
    		this.x = x;
    		this.cnt = cnt;
    	}
    }
}